0011-revise-e2e-test-framework.adoc

```diff
@@ -64,10 +63,356 @@ While the purpose of this spike was to figure out exactly what is wrong with the
+
+==== Other features of jest
+
+- Jest allows you to specify the number of expected assertions: https://jestjs.io/docs/en/expect#expectassertionsnumber
+
+=== Part 2
+
+==== The ideal test
+
+[source,js]
+----
+describe("happy path", () => {
+
+    beforeAll(async () => {
+        await ensureBitcoind();
+        await ensureParity();
+        await ensureCnd();
+        await ensureBtsieve();
+    });
+
+    test("bitcoin ether", () => {
+        let {alice, bob} = createActors();
+
+        alice.sendRequest("bitcoin", "ether");
+        bob.accept();
+        alice.fund();
+        bob.fund();
+        alice.redeem();
+        bob.redeem();
+
+        alice.assertSwapped();
+        bob.assertSwapped();
+    });
+});
+----
+
+==== The ideal test on steroids with `describe.each`
+
+From the documentation of Jest:
+
+> Use `describe.each` if you keep duplicating the same test suites with different data.
+`describe.each` allows you to write the test suite once and pass data in.
+
+[source,js]
+----
+describe.each([ ["bitcoin", "ether"], ["ether", "bitcoin"], ["bitcoin", "erc20"] ])('happy path',  (alpha, beta) => {
+    test(`${alpha} ${beta}`, () => {
+        let {alice, bob} = createActors();
+
+        alice.sendRequest({alpha, beta});
+        bob.accept();
+        alice.fund();
+        bob.fund();
+        alice.redeem();
+        bob.redeem();
+
+        alice.assertSwapped();
+        bob.assertSwapped();
+    });
+  },
+);
+----
+
+==== Philosophy
+
+The underlying idea of the test framework above can be summarized as: "smart defaults".
+
+Usually, a test will need two actors:
+Alice and Bob.
+Hence, the functions `ensureCnd` and `ensureBtsieve` will default to spawning two instances.
+If a different setup is needed, a list of required actors could be passed to those functions: `await ensureCnd(["alice", "bob", "charlie"])`
+Depending on the name of passed actor, the function `createActors` will return an object with properties according to those names.
+
+NOTE: This is probably impossible to implement in a fully generic way but we don't expect to have to scale much, 4-5 hardcoded actors names is probably fine.
+
+Similarily, all the "action"-functions have default parameters which could also be passed explicitly if a different behaviour is desired:
+
+[source,js]
+----
+let {alice, bob, charlie} = createActors();
+
+alice.sendRequest("bitcoin", "ether", {
+    to: charlie // <1>
+})
+----
+<1> Override, who we are sending the swap request to.
+
+[source,js]
+----
+let {alice, bob} = createActors();
+
+alice.sendRequest({beta: "bitcoin"}) // <1>
+----
+<1> We don't care about the alpha asset, will be replaced with a compatible default (not bitcoin).
+
+==== Output
+
+1. Use a proper logging framework instead of `console.log`
+2. Log useful stuff as part of the test execution: this is expected to grow and improve as we use the new test framework (i.e. add log statements and commit them if you debug a failing test).
+This includes:
```

Comment:
IMO passing tests should be quiet and failing tests noisy.  Or there should be flags to run the tests with different levels of verbosity.

Reply:
Absolutely. It might not be stated in the document, but we discussed that no logs should be emitted when a test passes. This may not be a worry if we decide to go for putting test logs in separate files, which means that stdout won't be cluttered with this information. In any case, this could be useful: https://github.com/rickhanlonii/jest-silent-reporter.

Reply:
Our assumption was to never log to stdout and hence always be quiet.
Different levels of verbosity only make sense if a failing test is reproducible, that is not always the case.

Hence I would recommend to always log on high verbosity but write it to a file so that stdout is clean. If you are interested in the log output, you can always check the file :)